### PR TITLE
Add Review Code

### PR DESCRIPTION
**Pull request by kaalu123**

### CODE REVIEW

File: Approval.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+This text file is written for getting approval of  xdufka1@fi.muni.cz on various aspects.
```
Comment (holubond) [inline Approval.txt:1]: **nitpick:** I'm not sure what is the exact purpose of this file, however, I believe approvals could be done using various different channels - file like this has nothing to do in similar projects.

File: .deepsource.toml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+version = 1
+
+[[analyzers]]
+name = "python"
+enabled = true
+
+  [analyzers.meta]
+  runtime_version = "3.x.x"
```
Comment (holubond) [inline .deepsource.toml:8]: **nitpick:** I don't think this file should be included in your main or release branch. Personally, I'd put this in a development or test branch as it is not an essential part of the resulting product... but it is just a subjective opinion and I might be completely wrong here :)

File: README.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
+**PA193-Project**
+
+This is a simple bech32m encoding and decoding tool written in python.
+
+The tool can be run from either CLI or STDIN
+
+
+**CLI**
+
+**Usage**: cli.py {enc,dec} hrp [-h] (-s S | -inf INF) [-iform {hex,bin,b64}] [-oform {hex,bin,b64}] [-of OF]
```
Comment (holubond) [inline README.md:10]: **issue:** I'm missing a python version here (especially due to the differences between python 2 and python 3).

File: bech32m.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+"""The function bech32_polymod(),bech32_hrp_expand(),bech32_verify_checksum(), bech32m_create_checksum() & convertbits()
+ have been taken from Reference implementation[1] and modified to cater for pure bech32m encoding/decoding only"""
```
Comment (holubond) [inline bech32m.py:24]: **thought:** Some parts are exact 1:1, including comments and badly-named variables.

File: bech32m.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    return [ord(x) >> 5 for x in hrp] + [0] + [ord(x) & 31 for x in hrp]
+
+
+def bech32_verify_checksum(hrp, data):
+    """Verify a checksum given HRP and converted data characters."""
+    const = bech32_polymod(bech32_hrp_expand(hrp) + data)
+    if const == BECH32M_CONST:
+        return 1
+    else:
+        return None
```
Comment (holubond) [inline bech32m.py:56]: **suggestion:** No need to have an `else` branch - you are making the code unnecessarily complex. Here it is not such a big issue, but in larger code chunks it will become very hard to read.

File: bech32m.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    return [(polymod >> 5 * (5 - i)) & 31 for i in range(6)]
+
+
+def convertbits(data, frombits, tobits, pad=True):
+    """General power-of-2 base conversion."""
+    acc = 0
+    bits = 0
+    ret = []
+    maxv = (1 << tobits) - 1
+    max_acc = (1 << (frombits + tobits - 1)) - 1
```
Comment (holubond) [inline bech32m.py:73]: **polish:** Variable names are not self-explanatory, which makes the code hard to read.

File: bech32m.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+        acc = ((acc << frombits) | value) & max_acc
+        bits += frombits
+        while bits >= tobits:
+            bits -= tobits
+            ret.append((acc >> bits) & maxv)
+    if pad:
+        if bits:
+            ret.append((acc << (tobits - bits)) & maxv)
+    elif bits >= frombits or ((acc << (tobits - bits)) & maxv):
+        return None
```
Comment (holubond) [inline bech32m.py:87]: **polish:** This block could be easily simplified - first evaluate whether you can return early and then potentially perform the appending.

File: bech32m.py [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+    elif len(check_str) % 4 != 0:
+        print("Invalid length of base 64 string. Check input string and input format")
+        return False
+
+    elif not all(x in base64_charset for x in check_str):
+        print("Non base64 Character Entered in String")
+        return False
+
+    else:
+        return True
```
Comment (holubond) [inline bech32m.py:111]: Again, unnecessary else blocks

File: bech32m.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+
+
+def check_if_str_is_base64(check_str):
+    """check if given string is valid base64 string"""
+
+    base64_charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="
+    if check_str == "":
+        return True
+
+    elif type(check_str) is not str:
```
Comment (holubond) [inline bech32m.py:98]: **polish:** Using `isinstance()` would make the code easier to read

File: bech32m.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+        print("Input given not a string. If input is raw bytes, select \"bin\" as input format")
+        return False
+
+    elif len(check_str) % 2 != 0:
+        print("Invalid Length of Hex String")
+        return False
+
+    elif not all(x in hex_charset for x in check_str):
```
Comment (holubond) [inline bech32m.py:133]: **thought:** This is possible to optimize using ascii-value comparisons. If done correctly (in a separate function), you would retain the same (if not better) level of readability.

File: bech32m.py [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+            return False
+        elif b != check_str:
+            return False
+        else:
+            return True
+
+    except AttributeError:
+        print("Invalid binary string")
+        return False
+
```
Comment (holubond) [inline bech32m.py:160]: **question:** ??? This seems like "automagic" to me... I cannot say whether the function is correct or not just from looking at it and I would require to see some tests that would give me higher confidence.

File: bech32m.py [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+            ret.append((acc >> bits) & maxv)
+    if pad:
+        if bits:
+            ret.append((acc << (tobits - bits)) & maxv)
+    elif bits >= frombits or ((acc << (tobits - bits)) & maxv):
+        return None
+    return ret
+
+
+def check_if_str_is_base64(check_str):
```
Comment (holubond) [inline bech32m.py:91]: **suggestion:** Not sure what the convention in Python is, but usually functions `check_*` work like `assert` - panic if the input is not expected. I'd recommend calling it e.g. `is_base64`, as functions prefixed with `is` return boolean value.

File: bech32m.py [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+    hex_charset = "ABCDEFabcdef0123456789"
+    if check_str == "":
+        return True
+
+    elif type(check_str) is not str:
+        print("Input given not a string. If input is raw bytes, select \"bin\" as input format")
+        return False
+
+    elif len(check_str) % 2 != 0:
+        print("Invalid Length of Hex String")
```
Comment (holubond) [inline bech32m.py:130]: **suggestion:** I don't recommend printing anything to the console in a non-ui layer of the application, it could cause hard-to-debug issues.

File: bech32m.py [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+
+    elif out_format == "hex":
+        return ascii_to_hex(bech32_str)
+
+    elif out_format == "b64":
+        return str_to_base64(bech32_str)
+
+    elif out_format == "bin":
+        raw_hex = ascii_to_hex(bech32_str)
+        return hex_to_raw(raw_hex)
```
Comment (holubond) [inline bech32m.py:269]: **suggestion:** Enum would make your life much easier

File: bech32m.py [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+        with open(input_file_name, mode) as f:
+            input_data = f.read()
+
+        if not input_data:
+            return ""
+        else:
+            return input_data
+
+    except IOError as e:
+        print("I/O error({0}): {1}".format(e.errno, e.strerror))
```
Comment (holubond) [inline bech32m.py:317]: **issue:** You are giving the user too much information, which could be potentially used to misuse your application.

File: bech32m.py [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+    """
+
+    if out_format == "hex":
+        return hex_str
+    elif out_format == "b64":
+        return hex_to_base64(hex_str)
+    elif out_format == "bin":
+        return hex_to_raw(hex_str)
+    else:
+        print("Invalid Output Format. Please Specify hex, b64 or bin")
```
Comment (holubond) [inline bech32m.py:291]: **question:** If you don't return anything, what will happen to the variable that is supposed to be assigned-to by this function? I assume it will be undefined and will cause the program to crash, if it is somehow used...

File: cli.py [hunk 17]
```diff
@@ -0,0 +1,166 @@
+import sys
+import argparse
+import bech32m
+import enc_dec
+
+
+def user_interface():
```
Comment (holubond) [inline cli.py:7]: **issue:** Function is waaay too long. It is therefore very hard to follow and is prone to errrors. Btw. by convention, functions should not be named using nouns - that's what classes are for.

File: cli.py [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+                #input_string = bytes(input_string, 'utf-8')
+                input_string = input_string.encode()
+
+        else:                                                          #if file name is provided read string from file
+            while not bech32m.read_file(infile, input_format):
+                infile = input("Enter input file name:")
+
+            input_string = bech32m.read_file(infile, input_format)
+
+        #validate input string with the input format provided, if string not in format ask for input string again
```
Comment (holubond) [inline cli.py:82]: **suggestion:** If you need to provide complex comments like this, you need to refactor.

File: enc_dec.py [hunk 19]
```diff
@@ -0,0 +1,249 @@
+import sys
+import bech32m
+from colorama import init, Fore, Style
```
Comment (holubond) [inline enc_dec.py:3]: **issue:** Colorama is not a part of the standard library

File: proj_test.py [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+             "122333",
+             "?",
+             "\/",
+             "/!@#$%^?/><~"
+             ]
+
+
+class Testenc_dec(unittest.TestCase):
+
+    def test_encode_pure_bech32m(self):
```
Comment (holubond) [inline proj_test.py:20]: **issue:** Split this into multiple functions to improve readability, I find it very hard to understand what you are testing.

File: cli.py [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+        if output_format is None:
+            output_format = "bech32m"
+
+        print("Mode:", mode, "HRP:", hrp, "Input String:", input_string, "Input Format:", input_format,
+              "Output Format:", output_format)
+
+        encoded_string = enc_dec.encode_pure_bech32m(hrp, input_string, input_format, output_format)
+
+        if not output_file:
+            print("\nEncoded String in {}".format(output_format[:].upper()), encoded_string)
```
Comment (holubond) [inline cli.py:137]: **suggestion:** Output like this makes it very hard to use this application by a different program

**Review by holubond — COMMENTED** [review]
Nice to see you used DeepSource, which was an extra to the project!

Thumbs up for the nice README & --help command!

I liked the fact you use colors in your terminal output, until I found out you use an external library, which was kinda prohibited.

I highly recommend you:
- Learn to work with git, including keeping your commits small and with descriptive messages. It is a must-have in any serious project
- Keep your functions small and simple. Do a bit of refactoring after you finish writing the code. The code is currently very hard to read and almost impossible to maintain in the long run (how long would it take you if I told you to add another output format?).
- Modularize. Having large files makes them hard to read.
- Listen to your tools. You used quite a lot of analyzers, however did not bother to fix the issues (or at least not all of them). Even the formatting issues are important, because if the code is not easy to follow, it would likely be rejected in the production (hard to follow = hard to maintain = expensive to have in the codebase, prone to issues).

Besides that, good job! I completely understand this was a difficult project and you had to overcome many challenges getting it to this state.

File: cli.py [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+        outfile = input("Enter output file name (press enter to print output to \"stdout\") : ")
+
+        return mode, hrp, input_string, infile, input_format, output_format, outfile
+
+
+def main():
+    mode, hrp, input_string, input_file, input_format, output_format, output_file = user_interface()
+
+    if input_string is None:
+        if not bech32m.read_file(input_file, input_format):
```
Comment (xmaslo) [inline cli.py:120]: issue (blocking): This call crashes application if non utf-8 character is in a file.

File: bech32m.py [hunk 23]
```diff
@@ -0,0 +1,341 @@
...
+    """
+
+    try:
+
+        if input_format == "bin":
+            mode = "rb"
+        else:
+            mode = "r"
+        with open(input_file_name, mode) as f:
+            input_data = f.read()
```
Comment (xmaslo) [inline bech32m.py:309]: issue (blocking): This call crashes application if non utf-8 character is in a file.

Comment (xmaslo) [inline bech32m.py:309]: issue (blocking): The `f.read()` with "r" mode crashes the application if a non utf-8 character is in a file.

Comment (xmaslo) [inline bech32m.py:309]: suggestion: Read a file as binary instead.

Comment (xmaslo) [inline bech32m.py:309]: **suggestion:** Read a file as binary instead.

File: bech32m.py [hunk 24]
```diff
@@ -0,0 +1,341 @@
...
+
+
+def check_if_str_is_bin(check_str):
+    """
+    check if input string has raw bytes
+    :param check_str: string to be checked
+    :return: true if string is raw bytes else false
+    """
+    try:
+        b = (check_str.decode()).encode()
```
Comment (xmaslo) [inline bech32m.py:148]: issue (blocking): If a string contains a non utf-8 character, the `decode()` crashes.

File: bech32m.py (hunk 23 above)
Comment (xmaslo) [inline bech32m.py:309]: **issue (blocking):** The `f.read()` with "r" mode crashes the application if a non-utf-8 character is in a file.

Example file: see review (files cannot be added in comments for some reason).

Comment (xmaslo) [inline bech32m.py:309]: **suggestion:** Read a file as binary instead.

File: bech32m.py (hunk 24 above)
Comment (xmaslo) [inline bech32m.py:148]: **issue (blocking):** If a string contains a non-utf-8 character, the `decode()` crashes.

Example file: see review (files cannot be added in comments for some reason).

File: README.md (hunk 3 above)
Comment (kaalu123) [reply README.md:10]: Python3

File: bech32m.py (hunk 8 above)
Comment (kaalu123) [reply bech32m.py:111]: A few tools reported issues if else block was removed

File: bech32m.py (hunk 15 above)
Comment (kaalu123) [reply bech32m.py:317]: I don't understand this completely.

**Review by xmaslo — COMMENTED** [review]
I mainly focused on dependency checking and fuzzing.

# Dependency checking
I used the **safety** for dependency checking. Unsurprisingly, no issues were found since the only standard library was allowed to be used. Just be careful next time since you used the **colorama** which is a violation of the assignment.

# Fuzzing
With fuzzing, I focused on the CLI and the bech32m.

### CLI
Whenever we provide the application with invalid command-line arguments, it crashes. Is this the desired behavior? Shouldn't we handle it differently?

### Bech32m
The main problem you had with your application is that you did not expect input other than utf-8. I detected these issues at two places and they are part of the comments. Here I also provide the example files and respective commands that crash the application (I couldn't add it as the comment directly because some sort of an error occurred).

`python3 cli.py enc bc -iform bin -inf fuzzing/e_bin_crashes/fuzz_5.txt`
[fuzz_5.txt](https://github.com/kaalu123/PA193-Project/files/8633766/fuzz_5.txt)

`python3 cli.py dec bc -inf fuzzing/d_crashes/fuzz_72.txt`
[fuzz_72.txt](https://github.com/kaalu123/PA193-Project/files/8633733/fuzz_72.txt)

# Other issues

- Why is it necessary to provide the **hrp** string when decoding? It should not be required.
- Suggestion: I would remove the **help.txt** file. It seems to be outdated (it mentions the main.py file which is not even part of the project) and there is already the **-h** command-line argument provided by the application.

# Conclusion
Overall, the implementation seems to be working as intended except for the above-mentioned issue. What I like about the application is the command-line interface which can use the arbitrary ordering of arguments and can be used inline as well as take input from stdin.

File: bech32m.py (hunk 15 above)
Comment (holubond) [reply bech32m.py:317]: I don't have much experience with python, but I believe `e.sterror` could contain for example name of the file that you are trying to open. If your application was used somewhere deep on a server and this error message got accidentally propagated to the client (a potential hacker), they might gain context of what's happening on the server and use it for their malicious attempts. Situations like this were discussed during lectures.

File: bech32m.py [hunk 25]
```diff
@@ -0,0 +1,341 @@
...
+    """Verify a checksum given HRP and converted data characters."""
+    const = bech32_polymod(bech32_hrp_expand(hrp) + data)
+    if const == BECH32M_CONST:
+        return 1
+    else:
+        return None
+
+
+def bech32m_create_checksum(hrp, data):
+    """Compute the checksum values given HRP and data."""
```
Comment (Milda00) [inline bech32m.py:60]: polish (documentation): The docstrings from the reference implementation could be unified with rest of the code to also describe function parameters.

File: bech32m.py (hunk 14 above)
Comment (Milda00) [reply bech32m.py:269]: nitpick (style): There seem to be two styles of formatting used for the if-elif constructs (lines between statements)

File: cli.py [hunk 26]
```diff
@@ -0,0 +1,166 @@
...
+        my_parser.add_argument('-of', type=str, help='output file path to write encoded/decoded string')
+        args = my_parser.parse_args()
+
+        if args.s is not None and args.iform == "bin":
+            args_s = args.s
+            args_str = args_s.encode()
+        else:
+            args_str = args.s
+
+        return args.m, args.hrp, args_str, args.inf, args.iform, args.oform, args.of
```
Comment (Milda00) [inline cli.py:31]: suggestion (non-blocking): This is a lot of related values, maybe return them in a single object.

File: cli.py [hunk 27]
```diff
@@ -0,0 +1,166 @@
...
+        group = my_parser.add_mutually_exclusive_group(required=True)
+        group.add_argument('-s', type=str, help='input string to be encoded/decoded')
+        group.add_argument('-inf', type=str, help='input file to read string to be encoded/decoded')
+        my_parser.add_argument('-iform', type=str, choices=['hex', 'bin', 'b64'], help='input format')
+        my_parser.add_argument('-oform', type=str, choices=['hex', 'bin', 'b64'], help='output format')
+        my_parser.add_argument('-of', type=str, help='output file path to write encoded/decoded string')
+        args = my_parser.parse_args()
+
+        if args.s is not None and args.iform == "bin":
+            args_s = args.s
```
Comment (Milda00) [inline cli.py:26]: question (dead code): What is this for?

File: cli.py [hunk 28]
```diff
@@ -0,0 +1,166 @@
...
+        hrp = input("Enter HRP[valid ascii]:")
+        while not enc_dec.validate_hrp(hrp):
+            hrp = input("Enter HRP[valid ascii]:")
+
+        #Enter and validate input format
+        input_format = input("Enter Input Format[b64/bin/hex]"
+                             "(Press Enter for Default[\"hex\" for \"enc\" and \"bech32m\" for \"dec\"]):")
+        while input_format not in ["b64", "bin", "hex", ""]:
+            input_format = input("Enter Valid Input Format[b64/bin/hex]"
+                                 "(Press Enter for Default[\"hex\" for \"enc\" and \"bech32m\" for \"dec\"]):")
```
Comment (Milda00) [inline cli.py:53]: issue (maintainability): There are repeated prompts in the code, I suggest either using a special function for getting user inputs, or at least parametrizing the messages and changing them only in one place.

File: cli.py [hunk 29]
```diff
@@ -0,0 +1,166 @@
...
+            input_string = input("Enter the input string:")
+            if not input_string and mode == "enc":
+                print("Null String Entered")
+            else:
+                while not input_string and mode == "dec":
+                    print("Empty String Cannot be decoded")
+                    input_string = input("Enter the input string:")
+
+            if input_format == "bin":
+                #input_string = bytes(input_string, 'utf-8')
```
Comment (Milda00) [inline cli.py:73]: question (dead code): What is this for?

File: cli.py [hunk 30]
```diff
@@ -0,0 +1,166 @@
...
+                input_string = input("Enter Valid \"base64\" input string:")
+
+        if input_format == "hex":
+            while not bech32m.check_if_str_is_hex(input_string):
+                input_string = input("Enter valid \"Hex\" input string:")
+
+        if input_format == "bin":
+            while not bech32m.check_if_str_is_bin(input_string):
+                input_string_s = input("Enter valid \"bin\" input string:")
+                #input_string = bytes(input_string_s, 'utf-8')
```
Comment (Milda00) [inline cli.py:94]: question (dead code): What is this for?

File: enc_dec.py [hunk 31]
```diff
@@ -0,0 +1,249 @@
...
+
+                return validated_str
+            else:
+                print("Invalid bin String")
+                return False
+        else:
+            print("Invalid Input Format!!!. Input string should be in \"hex\", \"b64\" or \"bin\"")
+            return None
+    else:
+        sys.exit(0)
```
Comment (Milda00) [inline enc_dec.py:73]: issue (logic): The code seems inconsistent - if validate_hrp fails, the program is killed here, in case of invalid input format, None is returned and the program is killed elsewhere.

File: enc_dec.py [hunk 32]
```diff
@@ -0,0 +1,249 @@
...
+            return True
+
+
+def validate_str_for_encoding(hrp, input_str, input_format):
+    """
+    validates a string for encoding and return validated string
+    :param hrp: string to specify Human-readable part
+    :param input_str: string to be encoded
+    :param input_format: input format of string to be encoded
+    :return: raw bytes of validated string
```
Comment (Milda00) [inline enc_dec.py:39]: issue (documentation): 4/7 return statements return either False or None, not raw bytes of validated string.

File: enc_dec.py [hunk 33]
```diff
@@ -0,0 +1,249 @@
...
+        elif input_format == "bin":
+            if bech32m.check_if_str_is_bin(input_str) or input_str == "":
+                validated_str = input_str
+
+                return validated_str
+            else:
+                print("Invalid bin String")
+                return False
+        else:
+            print("Invalid Input Format!!!. Input string should be in \"hex\", \"b64\" or \"bin\"")
```
Comment (Milda00) [inline enc_dec.py:70]: nitpick (output) - These messages should probably be printed to stderr

**Review by Milda00 — COMMENTED** [review]
I like the inclusion of README and help, the biggest issue that jumped out to me was all the tests clumped into one.
